### PR TITLE
Vinai/dolt merge Part 1. Just Fast Forward Merge

### PR DESCRIPTION
**Pull request by VinaiRachakonda**
This pr starts the Dolt_Merge process that works with a fast forward.

### CODE REVIEW

**Comment by VinaiRachakonda** [comment]
#benchmark

File: go/libraries/doltcore/sqle/database.go [hunk 1]
```diff
@@ -478,6 +478,25 @@ func (db Database) SetRoot(ctx *sql.Context, newRoot *doltdb.RootValue) error {
 	dsess := DSessFromSess(ctx.Session)
 	dsess.dbRoots[db.name] = dbRoot{hashStr, newRoot}
 
+	_, value := ctx.Get(sql.AutoCommitSessionVar)
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/database.go:481]: Definitely do not want this logic in here, not even temporarily. Should be handled by a layer above.

**Comment by coffeegoddd** [comment]
#benchmark

**Review by zachmu — COMMENTED** [review]
Looks like a good start except for the repo state management

**Comment by VinaiRachakonda** [comment]
#benchmark

**Review by zachmu — COMMENTED** [review]
Performance regression is probably due to you adding repo state updates to batch flushing. Think about ways to avoid doing this.

**Comment by VinaiRachakonda** [comment]
#benchmark

File: go/cmd/dolt/commands/sql.go [hunk 2]
```diff
@@ -1064,6 +1114,37 @@ func foundSubquery(node sqlparser.SQLNode) bool {
...
+		return false, nil
+	}
+}
+
+func hasDoltSQLFunction(node sqlparser.SQLNode) bool {
+	has := false
+	_ = sqlparser.Walk(func(node sqlparser.SQLNode) (keepGoing bool, err error) {
+		if f, ok := node.(*sqlparser.FuncExpr); ok {
+			name := strings.ToLower(f.Name.String())
+			if strings.Contains(name, "dolt_") {
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go:1137]: HasPrefix

File: bats/sql-merge.bats [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+INSERT INTO test VALUES (3);
+SELECT DOLT_COMMIT('-a', '-m', 'this is a ff');
+SELECT DOLT_CHECKOUT('master');
+SQL
+    run dolt sql -q "SELECT DOLT_MERGE('feature-branch');"
+    [ $status -eq 0 ]
+
+    run dolt log -n 1
+    [ $status -eq 0 ]
+    [[ "$output" =~ "this is a ff" ]] || false
```
Comment (zachmu) [inline bats/sql-merge.bats:33]: Check the data too?

File: bats/sql-merge.bats [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    run dolt sql -q "SELECT COUNT(*) FROM dolt_log"
+    [ $status -eq 0 ]
+    [[ "$output" =~ "3" ]] || false
+}
+
+@test "DOLT_MERGE correctly returns head and working session variables." {
```
Comment (zachmu) [inline bats/sql-merge.bats:40]: Need at least one test of error conditions

Comment (zachmu) [reply bats/sql-merge.bats:40]: And a test of the --no-ff option

File: go/cmd/dolt/commands/sql.go [hunk 5]
```diff
@@ -1064,6 +1114,37 @@ func foundSubquery(node sqlparser.SQLNode) bool {
 	return has
 }
 
+// hasDoltSQLFunction checks if a function is a dolt SQL function as defined in the dfunc package.
+func checkForDoltSQLFunction(query string) (bool, error) {
+	statement, err := sqlparser.Parse(query)
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go:1119]: No need to parse again, you've got the statement at the callsite

Comment (zachmu) [inline go/cmd/dolt/commands/sql.go:1119]: No need to parse again, you've got the statement at the callsite

File: go/libraries/doltcore/sqle/dfunctions/dolt_merge.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	if canFF {
+		err = executeFFMerge(ctx, false, dbData, cm)
+		if err != nil {
+			return nil, err
+		}
+		return cmh.String(), err
+	} else {
+		return nil, errors.New("Cannot handle this case yet.")
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dfunctions/dolt_merge.go:101]: Better error message

File: go/libraries/doltcore/sqle/dfunctions/dolt_merge.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	rv, err := cm2.GetRootValue()
+
+	if err != nil {
+		return errors.New("Failed to return root value.")
+	}
+
+	stagedHash, err := dbData.Ddb.WriteRootValue(ctx, rv)
+
+	if err != nil {
+		return errors.New("Failed to write database.")
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dfunctions/dolt_merge.go:137]: Just pass this along as is, more useful

File: go/libraries/doltcore/sqle/dfunctions/dolt_merge.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	workingHash := stagedHash
+	if !squash {
+		err = dbData.Ddb.FastForward(ctx, dbData.Rsr.CWBHeadRef(), cm2)
+
+		if err != nil {
+			return errors.New("Failed to write database")
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dfunctions/dolt_merge.go:145]: Same comment here

File: go/libraries/doltcore/sqle/database.go [hunk 9]
```diff
@@ -478,6 +478,11 @@ func (db Database) SetRoot(ctx *sql.Context, newRoot *doltdb.RootValue) error {
 	dsess := DSessFromSess(ctx.Session)
 	dsess.dbRoots[db.name] = dbRoot{hashStr, newRoot}
 
+	_, err = dsess.dbDatas[db.name].Ddb.WriteRootValue(ctx, newRoot)
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/sqle/database.go:481]: You are already writing root values explicitly in the action itself

**Review by zachmu — COMMENTED** [review]
Looks pretty good, needs a little work

File: bats/sql-merge.bats (hunk 4 above)
Comment (VinaiRachakonda) [reply bats/sql-merge.bats:40]: Did not implement no-ff yet

File: go/cmd/dolt/commands/sql.go [hunk 10]
```diff
@@ -948,6 +986,18 @@ func processNonInsertBatchQuery(ctx *sql.Context, se *sqlEngine, query string, s
 		return err
 	}
 
+	foundDoltSQLFunc, err := checkForDoltSQLFunction(query)
+	if err != nil {
+		return err
+	}
+
+	if foundDoltSQLFunc {
+		err = updateRepoState(ctx, se)
```
[the remark this answers is not in the document]
Comment (VinaiRachakonda) [reply go/cmd/dolt/commands/sql.go:995]: Actually think you should it before right? 

Say you do an INSERT and then do a dolt function. You want to do it first right?

**Comment by VinaiRachakonda** [comment]
#benchmark

File: go/cmd/dolt/commands/sql.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zachmu) [reply go/cmd/dolt/commands/sql.go:1002]: Yeah, you are right!

You should add a comment about why this is necessary, e.g.

"Some DOLT_* functions, like DOLT_MERGE, need to read an updated repo state to function correctly"

File: go/cmd/dolt/commands/sql.go [hunk 11]
```diff
@@ -948,6 +993,18 @@ func processNonInsertBatchQuery(ctx *sql.Context, se *sqlEngine, query string, s
 		return err
 	}
 
+	foundDoltSQLFunc, err := checkForDoltSQLFunction(sqlStatement)
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go:996]: This check needs to go in canProcessAsBatchInsert as well. If you did `INSERT INTO my_log (x) select dolt_merge(...)` you would need to flush there.

Also add a bats tests for DOLT_MERGE with non-insert statements

**Review by zachmu — APPROVED** [review]
Looks good! I had to check this out myself and experiment to convince myself it was right, so nice work getting it correct.

I have a feeling we're gonna be fooling around with this a lot more before all is said and done. So I'm not going to worry too much about getting this exactly right yet. But it does make me a little worried how much state management is happening here, in an uncoordinated fashion. I'll be thinking of ways to make this more manageable.

File: go/cmd/dolt/commands/sql.go (hunk 11 above)
Comment (VinaiRachakonda) [reply go/cmd/dolt/commands/sql.go:996]: Subqueries automatically get sent as nonBatched so I think the solution is to update the checkForDolSQLFunction to look at subqueries as well.

**Pull request merged**